### PR TITLE
feat(spanner): add `query_optimizer_statistics_package` support

### PR DESCRIPTION
**Pull request by skuruppu**
The optimizer statistics package can be set through `QueryOptions`, which can be configured through the following mechanisms.
1. At the `SpannerConnection` level.
1. Through the `SPANNER_OPTIMIZER_STATISTICS_PACKAGE` environment variable.
1. At a query level.
    
If the options are configured through multiple mechanisms then:
1. Options set at an environment variable level will override options configured at the `SpannerConnection` level.
1. Options set at a query-level will override options set at either the `SpannerConnection` or environment variable level.
    
If no options are set, the optimizer statistics package will default to:
1. The package the database is pinned to.
1. If the database is not pinned to a specific package, then the Cloud Spanner backend will use the "latest" version.

### CODE REVIEW

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.IntegrationTests/QueryOptionsTests.cs
Comment (jskeet) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.IntegrationTests/QueryOptionsTests.cs:43]:
```suggestion
                connection.QueryOptions = QueryOptions.Empty
                    .WithOptimizerVersion("1")
                    .WithOptimizerStatisticsPackage("auto_20191128_14_47_22UTC");
```

(I think we should use this fluent style ourselves, and encourage it in any samples etc.)

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.IntegrationTests/QueryOptionsTests.cs
Comment (jskeet) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.IntegrationTests/QueryOptionsTests.cs:65]: Ditto.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerCommandTests.cs
Comment (jskeet) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerCommandTests.cs:103]: And again. (I'm noting these just so you don't need to search for them.)

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerCommandTests.cs
Comment (jskeet) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerCommandTests.cs:123]: And here.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerCommandTests.cs
Comment (jskeet) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerCommandTests.cs:129]: This bit is in the other PR I've just approved, so should disappear after merge and rebase, yes?

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerCommandTests.cs
Comment (jskeet) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerCommandTests.cs:156]: And here.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerCommandTests.cs
Comment (jskeet) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerCommandTests.cs:195]: Here, and four lines later.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerCommandTests.cs
Comment (jskeet) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerCommandTests.cs:203]: Having seen this bit of code (with the validation later) three times now, I suspect a helper method would be useful.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/QueryOptions.cs
Comment (jskeet) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/QueryOptions.cs:34]: I'm not sure why I didn't spot this before (for OptimizerVersion) but this can be simplified to:

```csharp
 public string OptimizerStatisticsPackage => Proto.OptimizerStatisticsPackage;
```

(And that's the more idiomatic representation.)

**Review by jskeet — APPROVED** [review]
Generally fine - most comments are basically a single suggestion repeated many times :)

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerCommandTests.cs
Comment (skuruppu) [reply apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerCommandTests.cs:129]: Ah not quite. In #5292, I explicitly didn't modify all of these instances because I was already making the change in this PR. I only made changes to the emulator tests and not to the query options tests. These will stay in this PR (unless I should move this work to a separate PR?).

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/QueryOptions.cs
Comment (skuruppu) [reply apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/QueryOptions.cs:34]: Much nicer. Done.

**Comment by skuruppu** [comment]
All done, thanks @jskeet, your suggestions really improved the readability of the code.

**Comment by jskeet** [comment]
Great - I'm happy for you to rebase and merge unless there's anything you'd specifically like me to look at again. (I don't think any of my suggested changes would have been error-prone.)

**Comment by skuruppu** [comment]
I can't quite merge yet because the protos haven't been published :( I'm hoping to find out next week when we can publish those and hopefully soon after, I can merge it in. The expectation is by the end of the month.

**Comment by skuruppu** [comment]
Closing this PR for now since it'll be a little while before we'd merge this in.

**Pull request closed without merging**
